### PR TITLE
[Feature] Implemented Datetime object support

### PR DESCRIPTION
**Pull request by MannarAmuthan**
# Implemented Datetime object

Resolves: #647 

### What's Changed:
I didn't removed old contracts yet, means there is no any breaking changes, confirmed with tests. Added instantiation for datetime objects, once we agreed on these contracts, other utility methods for this datetime objects can be implemented on top of it. 

### Type of Change:

- [ ] Bug fix
- [x ] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#

### Housekeeping:

- [ x] Tests have been updated to reflect the changes done within this PR (if applicable).
- [ x] Documentation has been updated to reflect the changes done within this PR (if applicable).

#

#### Constructors
1. Datetime.new() -> Datetime
2. Datetime.new(String, String) -> Datetime
3. Datetime.new(Number) -> Datetime
#### Methods
1. datetimeObj.strptime() -> Number
2. datetimeObj.strftime(String) -> String

### CODE REVIEW

**Comment by MannarAmuthan** [comment]
Looks like there is some issue in handling gmt, will fix it

File: src/optionals/datetime.c [hunk 1]
```diff
@@ -163,8 +397,14 @@ Value createDatetimeModule(DictuVM *vm) {
...
      * Define Datetime methods
      */
+    
+    defineNative(vm, &module->values, "new", newDatetimeNative);
+    defineNative(vm, &module->values, "newUTC", newUTCDatetimeNative);
     defineNative(vm, &module->values, "now", nowNative);
     defineNative(vm, &module->values, "nowUTC", nowUTCNative);
+
+    // defineNative(vm, &module->values, "now", nowNative);
+    // defineNative(vm, &module->values, "nowUTC", nowUTCNative);
```
Comment (Jason2605) [inline src/optionals/datetime.c:407]: If these arent needed we can remove

```suggestion

```

File: tests/datetime/constants.du [hunk 2]
```diff
@@ -15,30 +15,30 @@ class TestDatetimeConstants < UnitTest {
     const DATE_TIME_FORMAT = "%a %b %d %H:%M:%S %Y";
 
     testSecondsInMinuteConstant() {
-        const startSec = Datetime.strptime(this.DATE_TIME_FORMAT, "Fri May 29 03:12:32 2020");
+        const startSec = Datetime.new(this.DATE_TIME_FORMAT, "Fri May 29 03:12:32 2020").strptime();
```
Comment (Jason2605) [inline tests/datetime/constants.du:18]: For these can we leave the existing tests alone but also add in new tests?

**Review by Jason2605 — COMMENTED** [review]
Couple of minor points

**Comment by Jason2605** [comment]
Appreciate the work you've done on this so i've stuck the hacktoberfest label on it (if you're participating) until I get some time to thoroughly check it and get it merged in!

**Comment by Jason2605** [comment]
I think I'm going to work on this PR (unless you want to continue with it) so that all of the object methods return a Datetime object rather than primitive values and have extra methods on the datetime object to convert to string / numeric values.

Means as we add more methods we can much easier chain things (e.g like addHour or whatever)

**Comment by MannarAmuthan** [comment]
Certainly, could you provide some contract ideas that I can attempt to implement ?.
Because current methods such as datetimeObj.strptime,  datetimeObj.strftime could not return Datetime object. Am I right?

**Comment by Jason2605** [comment]
Exactly yeah, we'd leave the current ones in and they can return the primitive values. Your new object methods would instead return a Datetime object, so:

```
Datetime.new(<values>); -> Datetime
Datetime.strptime(<string>, <string>); -> Datetime
Datetime.getTime(); -> Number
Datetime.strftime(<string>); -> String
Datetime.toString(); -> String (thinking do we need this if we have strftime actually? we could have this be strftime with no format specifier passed to it)


// Later down the line
Datetime.addMinutes(<number>); -> Datetime
Datetime.addHours(<number>); -> Datetime
...
```

**Comment by MannarAmuthan** [comment]
Thanks for Suggestions,
I assume that we have two different types of methods/functions, one is on Datetime module, and another is on Datetime object itself.

Methods of Datetime module
=====================
Already there
--------------
Datetime.new(); -> Datetime(Object) , current time object
Datetime.newUTC(); -> Datetime(Object) , current time object

To implement/modify
--------------
Datetime.new(<values>); -> Datetime (Object)
Does values means day,month,year, hour, etc ..?
Datetime.strptime(<string>, <string>); -> Datetime(Object) , currently it returns string i guess

Methods of Datetime Object
=====================
Already there
--------------
dateTime(object).strftime(<string>)  -> String , returns string

In future
---------
dateTime(object).addMinutes(<number>); -> Datetime(Object)
dateTime(object).addHours(<number>); -> Datetime(Object)
...

**Comment by Jason2605** [comment]
Apologies yeah, step time to return a date time object rather than a number, and then getTime to return a number

I had <values> because at the moment this is overloaded, that’s fine by me and can stay as is

**Comment by Jason2605** [comment]
No rush or anything @MannarAmuthan but just checking if you're still working on this?

**Comment by MannarAmuthan** [comment]
Hi yes , I am planning to work on this in weekend

**Comment by MannarAmuthan** [comment]
Hi @Jason2605 , Please look into it, and provide any suggestions or changes needed.

**Comment by Jason2605** [comment]
Will take a look later today, thanks @MannarAmuthan

File: docs/docs/standard-lib/datetime.md [hunk 3]
```diff
@@ -82,28 +114,31 @@ Datetime.now(); // Fri May 29 02:12:32 2020
...
-the first parameter being the date format, see [Datetime formats](#datetime-formats) and the second
-the date string in the format of the first parameter.
+Returns a user-defined datetime formatted string for a datetime object, see [Datetime formats](#datetime-formats).
+
+```cs
+const datetime = Datetime.new();
+datetime.strftime("Today is %A"); // Today is Friday
+```
+
+Returns in default format when user defined format is not provided
```
Comment (Jason2605) [inline docs/docs/standard-lib/datetime.md:137]: Be nice to note the default format here: `%a %b %d %H:%M:%S %Y`

File: src/optionals/datetime.c [hunk 4]
```diff
@@ -11,88 +11,78 @@
...
-        runtimeError(vm, "nowUTC() takes no arguments (%d given)", argCount);
-        return EMPTY_VAL;
-    }
+char *datetimeTypeToString(ObjAbstract *abstract) {
+    UNUSED(abstract);
 
-    time_t t = time(NULL);
-    struct tm tictoc;
-    char time[26];
+    char *queueString = malloc(sizeof(char) * 11);
```
Comment (Jason2605) [inline src/optionals/datetime.c:30]:
```suggestion
    char *datetimeString = malloc(sizeof(char) * 11);
```

Comment (Jason2605) [reply src/optionals/datetime.c:30]: *And same below

File: src/optionals/datetime.c [hunk 5]
```diff
@@ -11,88 +11,78 @@
...
 
-    // -1 to remove newline
-    return OBJ_VAL(copyString(vm, time, strlen(time) - 1));
-}
 
-static Value strftimeNative(DictuVM *vm, int argCount, Value *args) {
-    if (argCount != 1 && argCount != 2) {
-        runtimeError(vm, "strftime() takes 1 or 2 arguments (%d given)", argCount);
+    if (argCount > 1) {
+        runtimeError(vm, "strftime() takes 0 or 1 argument (%d given)", argCount);
```
Comment (Jason2605) [inline src/optionals/datetime.c:39]:
```suggestion
        runtimeError(vm, "strftime() takes 0 or 1 arguments (%d given)", argCount);
```

**Review by Jason2605 — COMMENTED** [review]
Thanks for this!! Few small comments really

**Comment by Jason2605** [comment]
@briandowns Be good to get your thoughts on this too

**Comment by briandowns** [comment]
I like this. 

I'm curious if we could simplify it though.

I was looking at working on a time module myself at one point and was going to model it after the Time package in Go. Essentially, the Time type is just a number and operations are performed on that value(s). There's a Duration type as well which is just an alias for an int64. The package's API is small and clean as well. https://cs.opensource.google/go/go/+/master:src/time/time.go;drc=bc2124dab14fa292e18df2937037d782f7868635;l=135

**Comment by MannarAmuthan** [comment]
Hi @briandowns , That one is great reference, Thanks for this !

I assume we are talking about the structure of datetime object , will take look into it.

File: src/optionals/datetime.c (hunk 5 above)
Comment (MannarAmuthan) [reply src/optionals/datetime.c]: These suggestions are done

**Comment by Jason2605** [comment]
> I like this.
> 
> I'm curious if we could simplify it though.
> 
> I was looking at working on a time module myself at one point and was going to model it after the Time package in Go. Essentially, the Time type is just a number and operations are performed on that value(s). There's a Duration type as well which is just an alias for an int64. The package's API is small and clean as well. https://cs.opensource.google/go/go/+/master:src/time/time.go;drc=bc2124dab14fa292e18df2937037d782f7868635;l=135

Yeah this is an interesting concept, will need to have a deeper look through the Go source but I think representing the time as a number internally makes a lot of sense to me

**Comment by MannarAmuthan** [comment]
Actually I read more on this, and found that classic **_time_t_** epoch integer, represents time after 1970. I changed the code, it is working fine, 

**New structure:**

typedef struct {
    long time;    /* timestamp integer*/
    int is_local; 
} Datetime;

But I worry, that we need the capability to create Datetime before 1970 also. 
If you have any thoughts on this let me know.

**Comment by briandowns** [comment]
As always, I'll defer to @Jason2605 but I think we can be fine with storing a wall time and mono time in int64_t values in a dateTime object. From there, that single object can be used for timers, time diffs, and we can cast up and down from time_t. In the case of timestamps prior to epoch, they end up being negative numbers.

**Comment by Jason2605** [comment]
> As always, I'll defer to @Jason2605 but I think we can be fine with storing a wall time and mono time in int64_t values in a dateTime object. From there, that single object can be used for timers, time diffs, and we can cast up and down from time_t. In the case of timestamps prior to epoch, they end up being negative numbers.

Honestly for this one, I'd probably go with whatever the go library is doing (still need to properly look at it sorry). A single numeric value storing the current time sounds perfect to me and going into negative numbers for dates prior to 1970 is fine (and what happens when you want a time before then when using unix time anyways yeah)

**Comment by briandowns** [comment]
A simple and naive assumption of how this might look:

```c
typedef struct {
    int64_t wall;
    int64_t mono;
} DateTime;

DateTime now() {
    struct timespec wall, mono;

    clock_gettime(CLOCK_MONOTONIC, &mono);
    clock_gettime(CLOCK_REALTIME, &wall);

    DateTime dt;
    dt.mono = mono.tv_sec;
    dt.wall = wall.tv_sec;

    return dt;
}
```

Rooted off of the DateTime class, we can any number of methods that will use the values therein. Timers using mono, time and date operations using wall.

**Comment by MannarAmuthan** [comment]
Thanks for all suggestions, I am planning to work on this , will provide updates

**Comment by MannarAmuthan** [comment]
Currently, I have created a datetime object to store time as an integer and included a flag for location (local or not). However, this is not quite complete; I need to implement ADD and SUB methods  ( and duration ) to ensure proper functionality. I will investigate this further and continue working on it.

**Comment by Jason2605** [comment]
Thanks for still working on this! Will be a really nice addition to get a more fully featured datetime library

**Comment by MannarAmuthan** [comment]
## Suggestion and Approaches for introducing timedelta (or) duration object:

**Naming:**  The name **_timedelta_** is inspired from python datetime package , meanwhile **_duration_** also makes sense

**Note:** These are just initial thoughts, welcom more suggestions, even hybrid approaches.

**To Remember:** 
In Dictu, there is no named or keyword arguments for function calls, which restrict us to defining clear contract for timedelta object.

### Approach 1: Constructing TimeDelta by dictionary

```
mydatetime = Datetime.new();

toAdd = Timedelta({
'days': 2,
'hours': 1
});  // Can have year , month, etc

mydatetime.add(toAdd); // Returns new datetime object
```

#### Pros

- Can have more operations such as getting delta difference of two datetime objects, as delta object, etc.
- Having clear difference with datetime object

#### Cons
- API might looks slightly complex, especially introducing new object.
- Constrcution of timedelta object looks little nasty


### Approach 2: Avoid delta object

```
mydatetime = Datetime.new();
mydatetime.addDays(1);
mydatetime.addMinutes(2);
mydatetime.addYears(1); 
//etc
```

or 

```
mydatetime = Datetime.new().addDays(1).addMinutes(2).subSeconds(1);
```

#### Pros

- Very simple and minimal in terms of design

#### Cons

- Limited functionality 
- Might lead to defining lot of methods

**Comment by Jason2605** [comment]
Thanks very much for the detailed write up! I personally thing the Timedelta option is the way to go - I feel like when dealing with calculating time differences it will be a lot easier to conditionally create a timedelta object than it will be to apply different methods - it also means we will be able to have an actual value representation of a length of time (we can then apply nice string methods to this for example)!

No kwargs is a valid gripe and would be a really nice addition, but as you say for now, a dictionary would be the next best replacement for instantiation

Be good to hear your thoughts too @briandowns

**Comment by Jason2605** [comment]
Yeah when you run `.add` or `.minus` (or whatever we decide the public API to be) I would expect a new datetime object yeah!

I'm not too fussed on Duration vs Timedelta either, Luxon a popular JS library seems to use Duration whereas Python's implementation as mentioned uses Timedelta, so open to either on the naming front.

Comparisons of objects will be an interesting one in the sense of would you expect the standard operators to work with these types (< > != ==) or would we have comparison methods (potentially opens up the question of user defined operator overloading too but can save that for another day)

**Comment by briandowns** [comment]
I'm not too concerned around naming Go uses an int64 as type Duration. Meh. 

As for object comparison, I'd probably at least expect a couple methods like t2.equals(t1);, and maybe a diff method.

**Comment by Jason2605** [comment]
So you'd prefer methods rather than overloading the operators?

Makes sense, yeah I think im siding towards `Duration` at the moment

**Comment by briandowns** [comment]
Yeah, for sure for methods over operators. Seems like the surface area for that is super small and still achieves the necessary functionality.

**Comment by briandowns** [comment]
If this has stagnated, I'd like to pick it up.

**Comment by Jason2605** [comment]
> If this has stagnated, I'd like to pick it up.

Have at it mate!

**Comment by MannarAmuthan** [comment]
Sorry for dropping this earlier, but I'd love to jump back in and help now. Let me know if there are any updates or changes needed from my side, and I'll get on it as soon as possible.

I guess, I've finished most of the things, but I am not sure about the failures. Need to check

**Comment by Jason2605** [comment]
Closing in favour of #750

**Pull request closed without merging**